core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchReaderIterator.java

```diff
@@ -497,26 +499,44 @@ private void doLookups(Map<String,Map<KeyExtent,List<Range>>> binnedRanges,
     for (final String tsLocation : locations) {
 
       final Map<KeyExtent,List<Range>> tabletsRanges = binnedRanges.get(tsLocation);
-      if (maxTabletsPerRequest == Integer.MAX_VALUE || tabletsRanges.size() == 1) {
-        QueryTask queryTask = new QueryTask(tsLocation, tabletsRanges, failures, receiver, columns);
-        queryTasks.add(queryTask);
+      if (options.isUseScanServer()) {
+        // Ignore the tablets location and find a scan server to use
+        ScanServerLocator ssl = context.getScanServerLocator();
+        tabletsRanges.forEach((k, v) -> {
+          try {
+            String location = ssl.reserveScanServer(new TabletIdImpl(k));
```

Comment:
This seems really expensive doing an RPC to ZK before a scan and an RPC to ZK after the scan to unreserve.  I was thinking we could possibly just select a scan server and try to use it and if its busy it would throw some sort of thrift exception that indicates its busy.  When the client gets the busy exception it will try another scan server, similar to the reservation failing here. However this does not require going to ZK and also the busy check on scan servers could probably be orders of magnitude faster than ZK as it would not require persisting data in the ZK edit log and getting ZK server side quorom for a write.

One way a busy check could work on a scan server is that it has a fixed number of threads in a thread pool using a [SynchronousQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/SynchronousQueue.html).  When a scan task is submitted to this thread pool and all threads are busy the thread pool should reject it.  When the thread pool rejection happens then we can turn around and throw the busy thrift exception. All of this should be really quick on the scan server.

Reply:
As currently implemented a ScanServer performs one scan at a time, much like the Compactor performs one compaction at a time. I figured that if a ScanServer had many threads and performed more than one scan at a time, then we would potentially run into the same situation we have in in the TabletServer w/r/t memory usage.

Reply:
>  I figured that if a ScanServer had many threads and performed more than one scan at a time, then we would potentially run into the same situation we have in in the TabletServer w/r/t memory usage.

I can see the benefit of a single thread.  We could start off implementing the thread pool with syncQ with a hard coded thread of size of one and get this really fast busy exception behavior.  Later on, as we gain experience, we could refine the scan server config and make what causes busy exceptions on a scan server configurable.

Reply:
Another thought w/r/t to the load balancer code and the number RPCs to ZK, we could move the logic into the Manager and reduce the ZK traffic with ZooCache. I'm not tied to the current approach, I was just trying to make a simple reservation system that could be replaced with a user supplied class that performs more complex logic.

> We could start off implementing the thread pool with syncQ with a hard coded thread of size of one and get this really fast busy exception behavior.

I'm good with that. The Scan Server already returns an error if its working on a scan for another client. I'm currently working on some more ITs and running down an issue I am seeing in one of them. Let's talk more about potential design changes.

Reply:
Not sure if it achievable but it would be nice if we could have information about scan servers from ZK that is cacheable on the client side (so clients do not have to go to ZK frequently), a busy signal from scan servers, and the previous two bits of information available to client side plugin that makes decisions about which scan servers to use.  If that is workable it could avoid any single bottleneck processes in the cluster hopefully allowing this to scale up to many thousands of scan servers and thousands of clients.

Reply:
My latest commit added more tests in ScanServerIT. I need to add a couple more tests that involve the BatchScanner, but I'm pretty well convinced that this approach will work. I think the ScanServer locator piece might be the most complex piece about this change when we are done with it. From a load balancer perspective, it would be nice to know the following for each ScanServer:

- last N extents scanned (increased probability of cache hit?).
- local TabletServers Tablets, if any (increased probability of locality or page cache hit?)

The extent information is likely sensitive and not something we want client-side. If we choose to use this information for load balancing, then it might have to go into the Manager.